File: Angular/src/app/api.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ApiService {
  private apiUrl = 'http://localhost:8000/api';

  constructor(private http: HttpClient) {}

  getHeaders() {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json'
    });
  }

   // Send chat message to chatbot
   sendChatMessage(message: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/chatbot`, { message }, { headers: this.getHeaders() });
  }

  // Budget
  getAllBudgets(): Observable<any> {
    return this.http.get(`${this.apiUrl}/budget/all`, { headers: this.getHeaders() });
  }

  getBudget(): Observable<any> {
    return this.http.get(`${this.apiUrl}/budget`, { headers: this.getHeaders() });
  }

  addBudget(data: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/budget`, data, { headers: this.getHeaders() });
  }

  updateBudget(data: any, budgetId: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/budget/${budgetId}`, data, { headers: this.getHeaders() });
  }

  deleteBudget(budgetId: any): Observable<any> {
    return this.http.delete(`${this.apiUrl}/budget/${budgetId}`, { headers: this.getHeaders() });
  }

  // Goals
  getAllGoals(): Observable<any> {
    return this.http.get(`${this.apiUrl}/goal/all`, { headers: this.getHeaders() });
  }

  getGoal(): Observable<any> {
    return this.http.get(`${this.apiUrl}/goal`, { headers: this.getHeaders() });
  }

  addGoal(data: any): Observable<any> {
      return this.http.post(`${this.apiUrl}/goal`, data, { headers: this.getHeaders() });
  }

  updateGoal(data: any, goalId: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/goal/${goalId}`, data, { headers: this.getHeaders() });
  }

  deleteGoal(goalId: any): Observable<any> {
    return this.http.delete(`${this.apiUrl}/goal/${goalId}`, { headers: this.getHeaders() });
  }

  // Expenses
  getAllExpenses(): Observable<any> {
    return this.http.get(`${this.apiUrl}/expenses/all`, { headers: this.getHeaders() });
  }

  getExpenses(): Observable<any> {
    return this.http.get(`${this.apiUrl}/expenses`, { headers: this.getHeaders() });
  }

  addExpense(data: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/expenses`, data, { headers: this.getHeaders() });
  }

  updateExpense(data: any, id: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/expenses/${id}`, data, { headers: this.getHeaders() });
  }

  deleteExpense(id: any): Observable<any> {
      return this.http.delete(`${this.apiUrl}/expenses/${id}`, { headers: this.getHeaders() });
  }

  analyzeExpenses(): Observable<any> {
    return this.http.get(`${this.apiUrl}/analyze-expenses`, { headers: this.getHeaders() });
  }

  // Categories
  getCategories(): Observable<any> {
    return this.http.get(`${this.apiUrl}/categories`, { headers: this.getHeaders() });
  }

  addCategory(data: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/categories`, data, { headers: this.getHeaders() });
  }

  updateCategory(data: any, id: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/categories/${id}`, data, { headers: this.getHeaders() });
  }

  deleteCategory(id: any, newCategory: string): Observable<any> {
    const body = { new_category: newCategory };
    return this.http.delete(`${this.apiUrl}/categories/${id}`, {
      headers: this.getHeaders(),
      body: body
    });
  }

  getCategorySuggestions(): Observable<any> {
    return this.http.get(`${this.apiUrl}/categories/suggest`, { headers: this.getHeaders() });
  }

  getCategoryLabels(): Observable<any> {
    return this.http.get(`${this.apiUrl}/categories/label`, { headers: this.getHeaders() });
  }
}

File: Flutter/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class ApiService {
  static final String _apiUrl = 'http://localhost:8000/api';

  // 'http://10.0.2.2:8000/api' Android emulator and make sure to run "php artisan serve --host=0.0.0.0 --port=8000"
  // 'http://localhost:8000/api' Web or other platforms

  Future<Map<String, String>> _getHeaders() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('token');
    return {
      'Authorization': 'Bearer $token',
      'Content-Type': 'application/json',
    };
  }

  // Helper method to handle responses
  dynamic _handleResponse(http.Response response) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Error: ${response.statusCode} - ${response.body}');
    }
  }

  // Send chat message to chatbot
  Future<dynamic> sendChatMessage(String message) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('$_apiUrl/chatbot'),
      headers: headers,
      body: jsonEncode({'message': message}),
    );
    return _handleResponse(response);
  }

  // Budget
  Future<dynamic> getAllBudgets() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/budget/all'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> getBudget() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/budget'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> addBudget(Map<String, dynamic> data) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('$_apiUrl/budget'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> updateBudget(Map<String, dynamic> data, int budgetId) async {
    final headers = await _getHeaders();
    final response = await http.put(
      Uri.parse('$_apiUrl/budget/$budgetId'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> deleteBudget(int budgetId) async {
    final headers = await _getHeaders();
    final response = await http.delete(
      Uri.parse('$_apiUrl/budget/$budgetId'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  // Goals
  Future<dynamic> getAllGoals() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/goal/all'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> getGoal() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/goal'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> addGoal(Map<String, dynamic> data) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('$_apiUrl/goal'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> updateGoal(Map<String, dynamic> data, int goalId) async {
    final headers = await _getHeaders();
    final response = await http.put(
      Uri.parse('$_apiUrl/goal/$goalId'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> deleteGoal(int goalId) async {
    final headers = await _getHeaders();
    final response = await http.delete(
      Uri.parse('$_apiUrl/goal/$goalId'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  // Expenses
  Future<dynamic> getAllExpenses({int page = 1, int perPage = 1}) async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/expenses/all?page=$page&per_page=$perPage'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> getExpenses() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/expenses'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> addExpense(Map<String, dynamic> data) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('$_apiUrl/expenses'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> updateExpense(Map<String, dynamic> data, int id) async {
    final headers = await _getHeaders();
    final response = await http.put(
      Uri.parse('$_apiUrl/expenses/$id'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> deleteExpense(int id) async {
    final headers = await _getHeaders();
    final response = await http.delete(
      Uri.parse('$_apiUrl/expenses/$id'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> analyzeExpenses() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/analyze-expenses'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  // Categories
  Future<dynamic> getCategories() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/categories'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> addCategory(Map<String, dynamic> data) async {
    final headers = await _getHeaders();
    final response = await http.post(
      Uri.parse('$_apiUrl/categories'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> updateCategory(Map<String, dynamic> data, int id) async {
    final headers = await _getHeaders();
    final response = await http.put(
      Uri.parse('$_apiUrl/categories/$id'),
      headers: headers,
      body: jsonEncode(data),
    );
    return _handleResponse(response);
  }

  Future<dynamic> deleteCategory(int id, String newCategory) async {
    final headers = await _getHeaders();
    final response = await http.delete(
      Uri.parse('$_apiUrl/categories/$id'),
      headers: headers,
      body: jsonEncode({'new_category': newCategory}),
    );
    return _handleResponse(response);
  }

  Future<dynamic> getCategorySuggestions() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/categories/suggest'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> getCategoryLabels() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/categories/label'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  // Stocks
  Future<dynamic> getEgyptStocks() async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/stocks/egypt'),
      headers: headers,
    );
    return _handleResponse(response);
  }

  Future<dynamic> getStockDetails(String symbol) async {
    final headers = await _getHeaders();
    final response = await http.get(
      Uri.parse('$_apiUrl/stocks/details/$symbol'),
      headers: headers,
    );
    return _handleResponse(response);
  }
}
